crates/c-api/src/ref.rs

```diff
@@ -43,3 +95,60 @@ pub extern "C" fn wasm_ref_set_host_info_with_finalizer(
     eprintln!("`wasm_ref_set_host_info_with_finalizer` is not implemented");
     std::process::abort();
 }
+
+type wasmtime_externref_finalizer_t = fn(*mut c_void);
+
+struct CExternRef {
+    data: *mut c_void,
+    finalizer: Option<wasmtime_externref_finalizer_t>,
+}
+
+impl Drop for CExternRef {
+    fn drop(&mut self) {
+        if let Some(f) = self.finalizer {
+            f(self.data);
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_externref_new(data: *mut c_void) -> wasm_val_t {
+    wasm_val_t::from_val(Val::ExternRef(Some(ExternRef::new(CExternRef {
+        data,
+        finalizer: None,
+    }))))
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_externref_new_with_finalizer(
+    data: *mut c_void,
+    finalizer: Option<wasmtime_externref_finalizer_t>,
+) -> wasm_val_t {
+    wasm_val_t::from_val(Val::ExternRef(Some(ExternRef::new(CExternRef {
+        data,
+        finalizer,
+    }))))
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_externref_data(val: &wasm_val_t, datap: *mut *mut c_void) -> bool {
```

Comment:
This I think could be `&mut *mut c_void` to improve safety a bit (and document at the type level that it cant be null